internal/jobcompleter/job_completer.go

```diff
@@ -56,125 +64,290 @@ type InlineJobCompleter struct {
+// the vast common case under normal operation. The completer embeds an
+// AsyncCompleter to perform other non-`completed` state completions.
+type BatchCompleter struct {
+	baseservice.BaseService
+	startstop.BaseStartStop
+	withSubscribe
+
+	asyncCompleter   *AsyncCompleter // used for non-complete completions
+	debounceChan     *chanutil.DebouncedChan
+	exec             PartialExecutor
+	ready            chan struct{}
+	setStateParams   map[int64]*batchCompleterSetState
+	setStateParamsMu sync.Mutex
+}
+
+func NewBatchCompleter(archetype *baseservice.Archetype, exec PartialExecutor) *BatchCompleter {
+	return baseservice.Init(archetype, &BatchCompleter{
+		asyncCompleter: NewAsyncCompleter(archetype, exec),
+		exec:           exec,
+		setStateParams: make(map[int64]*batchCompleterSetState),
+	})
+}
+
+func (c *BatchCompleter) Start(ctx context.Context) error {
+	stopCtx, shouldStart, stopped := c.StartInit(ctx)
+	if !shouldStart {
+		return nil
+	}
+
+	c.ready = make(chan struct{})
+
+	go func() {
+		// This defer should come first so that it's last out, thereby avoiding
+		// races.
+		defer close(stopped)
 
-		func() {
-			c.subscribeFuncMu.Lock()
-			defer c.subscribeFuncMu.Unlock()
+		c.Logger.InfoContext(ctx, c.Name+": Run loop started")
+		defer c.Logger.InfoContext(ctx, c.Name+": Run loop stopped")
 
-			if c.subscribeFunc != nil {
-				c.subscribeFunc(CompleterJobUpdated{Job: job, JobStats: stats})
+		c.debounceChan = chanutil.NewDebouncedChan(stopCtx, 100*time.Millisecond)
+
+		close(c.ready)
+
+		for {
+			select {
+			case <-stopCtx.Done():
+				// Try to insert last batch before leaving. Note we use the
+				// original context so operations aren't immediately cancelled.
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
+
+				return
+			case <-c.debounceChan.C():
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
 			}
-		}()
+		}
+	}()
+
+	return nil
+}
 
+func (c *BatchCompleter) handleBatch(ctx context.Context) error {
+	var setStateBatch map[int64]*batchCompleterSetState
+	func() {
+		c.setStateParamsMu.Lock()
+		defer c.setStateParamsMu.Unlock()
+
+		setStateBatch = c.setStateParams
+
+		// Don't bother resetting the map if there's nothing to process,
+		// allowing the completer to idle efficiently.
+		if len(setStateBatch) > 0 {
+			c.setStateParams = make(map[int64]*batchCompleterSetState)
+		} else {
+			// Set nil to avoid a data race below in case the map is set as a
+			// new job comes in.
+			setStateBatch = nil
+		}
+	}()
+
+	if len(setStateBatch) < 1 {
 		return nil
-	})
+	}
+
+	// Use a single `finalized_at` value for the whole batch. Not the greatest
+	// thing maybe, but makes things much easier.
+	var sampleFinalizedAt time.Time
+	for _, setState := range setStateBatch {
+		sampleFinalizedAt = *setState.Params.FinalizedAt
+		break
+	}
```

Comment:
Unfortunately I think this change is a pretty substantial degradation. With this PR we're now debouncing up to 100ms, meaning jobs could have up to 100ms of randomly distributed padding added to their `finalized_at`, making it so `attempted_at - finalized_at` is no longer representative of their duration (at least not for reasonably fast jobs).

I'm sure it will be a bit less efficient, but I think we'd be better off keeping a `finalized_at` for each job as reported at the time it finished executing. Then your `JobSetCompleteIfRunningManyParams` can be a list of two arrays: the IDs, and the corresponding timestamps for each.

I also wanted to take this moment to raise another question. As I've progressed on the UI, I don't really love the limitations of trying to deduce job statistics from the handful of available timestamps (`inserted_at`, `attempted_at`, `finalized_at`, and those in `errors`). It's too complicated and leaves odd undefined gaps. I can paper over some of those at the UI level, but it has held me back from some stuff I'd like to ship like a more reliable execution timeline.

I've started wondering whether it'd be better if we tracked some sort of job stats object that we could keep as an array on the `JobRow` and as a jsonb array in Postgres. Possibly it could be merged together with `errors` so there's a single large jsonb ala `executions` or `attempts`. It would allow for some nice built-in stats even without hooking up an external monitoring system to consume all job stats in real time. I'm not sure how this will impact performance in the happy path though, because you'd then always be setting/updating a jsonb TOAST value as well as setting a timestamp.

Thoughts on both these fronts?

Reply:
> Unfortunately I think this change is a pretty substantial degradation. With this PR we're now debouncing up to 100ms, meaning jobs could have up to 100ms of randomly distributed padding added to their `finalized_at`, making it so `attempted_at - finalized_at` is no longer representative of their duration (at least not for reasonably fast jobs).

Fair enough. I just pushed up a change that also sets `finalized_at` specifically for each job, and a new set of tests.

> I also wanted to take this moment to raise another question. As I've progressed on the UI, I don't really love the limitations of trying to deduce job statistics from the handful of available timestamps (`inserted_at`, `attempted_at`, `finalized_at`, and those in `errors`). It's too complicated and leaves odd undefined gaps. I can paper over some of those at the UI level, but it has held me back from some stuff I'd like to ship like a more reliable execution timeline.

OOC, if you excluded errors/snoozes, are there any particular timings you were thinking of that you can't calculate using the existing fields?

Seems plausible, although I'd definitely try top get away with either (1) consolidating to one jsonb field, or (2) stuff more stats into the `errors` field that we already have. Having to look at two fields that may have a different number of elements and try to match them up against each sounds awful.

I'll also give my typical response that jsonb isn't very ergonomic. It's harder to use in queries, it's more obfuscated in query results, and you don't get autocomplate on names in psql. So as I mentioned with putting cancels in metadata, IMO the less we have of it the better. We'd have to put stats for retries in jsonb, but if we could get away with it only for errors, that might be better.

Ditching the name `errors` should sure make things easier for use in Sequel/ActiveRecord in Ruby land though. The hacks needed to work around this naming are kind of crazy.